File: output/amazon_q/php/hard/q36-merge-k-sorted-lists.php
<?php
/** https://leetcode.com/problems/merge-k-sorted-lists */
// Problem: Merge K Sorted Lists

//Problem Description:
//You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
//Merge all the linked-lists into one sorted linked-list and return it.

// Constraints:
// k == lists.length
// 0 <= k <= 10^4
// 0 <= lists[i].length <= 500
// -10^4 <= lists[i][j] <= 10^4
// lists[i] is sorted in ascending order.
// The sum of lists[i].length will not exceed 10^4.


//Code Structure:

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode[] $lists
     * @return ListNode
     */
    function mergeKLists($lists) {
        $mergedList = null;
        $current = null;

        while (true) {
            $minIndex = -1;
            $minValue = PHP_INT_MAX;

            foreach ($lists as $index => $list) {
                if ($list !== null && $list->val < $minValue) {
                    $minIndex = $index;
                    $minValue = $list->val;
                }
            }

            if ($minIndex === -1) {
                break;
            }

            if ($mergedList === null) {
                $mergedList = new ListNode($minValue);
                $current = $mergedList;
            } else {
                $current->next = new ListNode($minValue);
                $current = $current->next;
            }

            $lists[$minIndex] = $lists[$minIndex]->next;
        }

        return $mergedList;
    }
}
    

File: output/amazon_q/java/hard/q36-merge-k-sorted-lists.java
/** https://leetcode.com/problems/merge-k-sorted-lists */
// Problem: Merge K Sorted Lists

//Problem Description:
//You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
//Merge all the linked-lists into one sorted linked-list and return it.

// Constraints:
// k == lists.length
// 0 <= k <= 10^4
// 0 <= lists[i].length <= 500
// -10^4 <= lists[i][j] <= 10^4
// lists[i] is sorted in ascending order.
// The sum of lists[i].length will not exceed 10^4.


//Code Structure:

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        while (lists.length > 1) {
            ListNode[] merged = new ListNode[(lists.length + 1) / 2];
            for (int i = 0; i < lists.length; i += 2) {
                merged[i / 2] = merge(lists[i], i + 1 < lists.length ? lists[i + 1] : null);
            }
            lists = merged;
        }
        return lists[0];
    }
    
    private ListNode merge(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0), curr = dummy;
        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                curr.next = l1;
                l1 = l1.next;
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        curr.next = l1 != null ? l1 : l2;
        return dummy.next;
    }
}
    